docs/authentication.md

```diff
@@ -0,0 +1,318 @@
+
+```jsx
+// pages/profile.js
+
+export default Profile = () => {
+  // Fetch the user client-side
+  const { user } = useUser({ redirectTo: '/login' });
+
+  // Server-render loading skeleton
+  if (!user) {
+    return <LoadingSkeleton />;
+  }
+
+  // Once the user request finishes, show the user's name
+  return <p>{user.name}</p>;
+};
+```
+
+### Server-side Rendering
+
+If you export an `async` function called `getServerSideProps` from a page, Next.js will pre-render this page on each request using the data returned by `getServerSideProps`.
+
+```jsx
+export async function getServerSideProps(context) {
+  return {
+    props: {} // Will be passed to the page component as props
+  };
+}
+```
+
+Let's transform the profile example from earlier to use server-side rendering. If there's a session, return `session` as a prop to the `Profile` component in the page. Otherwise, redirect to `/login` on the server-side. Notice there is not a loading skeleton in this example.
+
+```jsx
+// pages/profile.js
+
+export default Profile = ({ session }) => {
+  return <p>{session.user.name}</p>;
+};
+
+export async function getServerSideProps({ req, res }) {
+  // Get the user's session based on the request
+  const session = await getSession(req);
+
+  if (!session) {
+    // If no user, redirect to login
+    res.writeHead(307, { Location: '/login' });
+    res.end();
+    return { props: {} };
+  }
+
+  // If there is a user, return the current session
+  return { props: { session } };
+}
+```
+
+An advantage of this pattern is preventing a flash of unauthenticated content before redirecting. It's important to note fetching user data in `getServerSideProps` will block rendering until the request to your authentication provider resolves. To prevent creating a bottleneck and decreasing your TTFB ([Time to First Byte](https://web.dev/time-to-first-byte/)), you should ensure your authentication lookup is fast. Otherwise, consider static generation.
+
+There is an [open RFC](https://github.com/vercel/next.js/discussions/14890) to improve redirecting inside `getServerSideProps`.
+
+## Providers
+
+Now that we've discussed authentication patterns, let's look at specific providers and explore how they're used with Next.js.
+
+### Firebase
+
+When using [Firebase Authentication](/blog/nextjs-firebase-serverless), we recommend using the static generation pattern. There currently isn't a way to log-in server-side using the Firebase Admin SDK.
+
+It is possible to use the Firebase Client SDK to generate an ID token and forward it directly to Firebase's REST API on the server to log-in. However, requests to Firebase might take some time to resolve, depending on your user's location.
+
+Let's look an example creating a `useUser` hook. This hook allows you to fully control where you'll call `signinWithGitHub`, or any other authentication provider with Firebase.
+
+```jsx
+import React, { useState, useEffect, useContext, createContext } from 'react';
```

Comment:
I think this code sample can be confusing for many users. I think giving an overall view of how Firebase auth works with Next.js and then linking to the example should be fine.